pims/base_frames.py

```diff
@@ -528,3 +528,268 @@ def process(img_or_iterable, *args, **kwargs):
+                             'used.')
+        self.name = name
+        self._default = default
+        self.size = size
+        if aggregate and iterate:
+            raise ValueError('Dimensions cannot aggregate and be iterable '
+                             'simultaneously.')
+        self._aggregate = aggregate
+        self._iterate = iterate
+
+    @property
+    def size(self):
+        return self._size
+
+    @size.setter
+    def size(self, value):
+        if value <= 0:
+            raise ValueError('Dimension size should be greater than zero.')
+        self._size = int(value)
+        self.default = self._default  # recheck default value validity
+
+    @property
+    def default(self):
+        return self._default
+
+    @default.setter
+    def default(self, value):
+        if value >= self.size:
+            raise ValueError('Default value out of bounds')
+        self._default = value
+
+    @property
+    def aggregate(self):
+        return self._aggregate
+
+    @aggregate.setter
+    def aggregate(self, value):
+        self._aggregate = value
+        if value:  # iterate and aggregate cannot both be True
+            self._iterate = False
+
+    @property
+    def iterate(self):
+        return self._iterate
+
+    @iterate.setter
+    def iterate(self, value):
+        self._iterate = value
+        if value:  # iterate and aggregate cannot both be True
+            self._aggregate = False
+
+    def __repr__(self):
+        s = "<FramesDimension>; Name: {0}; Size: {1}; Status: "
+        s = s.format(self.name, self.size)
+        if self.iterate:
+            s += "iterate"
+        elif self.aggregate:
+            s += "aggregate"
+        else:
+            s += "defaults to {}".format(self.default)
+        return s
+
+
+class Multidimensional(FramesSequence):
+    """ A base class defining a FramesSequence with an arbitrary number of
+    dimensions. The properties `aggregate` and `iterate` define the functions
+    of each dimension.
+
+    Subclassed readers need to define `frame_shape_2D` and `get_frame_2D`.
+    Dimensions need to be instanciated using the available method `add_dim`.
+
+    Examples
+    --------
+    >>> class MDummy(Multidimensional):
+    ...    @property
+    ...    def pixel_type(self):
+    ...        return 'uint8'
+    ...    @property
+    ...    def frame_shape_2D(self):
+    ...        return self._frame_shape_2D
+    ...    def __init__(self, shape, **dims):
+    ...        for name in dims:
+    ...            self.add_dim(name, dims[name])
+    ...        self._frame_shape_2D = shape
+    ...    def get_frame_2D(self, **ind):
+    ...        return np.zeros(self.frame_shape_2D, dtype=self.pixel_type)
+
+    >>> frames = MDummy((64, 64), t=80, c=2, z=10, m=5)
+    >>> frames.aggregate = 'cz'
+    >>> frames.iterate = 't'
+    >>> frames.dims['m'].default = 3
+    >>> frames[5]  # returns Frame at T=5, M=3 with shape (2, 10, 64, 64)
+    """
+    def clear_dims(self):
```

Comment:
Is this and `add_dim` really useful as public functions?  I have not mad it farther down the code yet, but it seems that once the reader is created, the dimensions should not change.